File: adk/asc/weil-sdk/assembly/collections/vec.ts
import { WeilId } from './weil-id'
import { IndexOutOfBoundsError, WeilError } from '../error'
import { Memory } from '../runtime'

/**
 * A persistent vector collection that stores ordered elements in the Weil runtime state.
 * 
 * @template T - The type of elements in the vector
 */
export class WeilVec<T> {
  /** The unique identifier for this vector's state storage */
  stateId: WeilId
  /** The current length of the vector */
  length: usize

  /**
   * Creates a new WeilVec instance with the specified state ID.
   * 
   * @param stateId - The unique identifier for this vector's state storage
   */
  constructor(stateId: WeilId) {
    this.stateId = stateId
    this.length = 0
  }

  /**
   * Gets the base state path for this vector.
   * 
   * @returns The string representation of the state ID
   */
  private getBaseStatePath(): string {
    return this.stateId.toString()
  }

  /**
   * Generates a unique state tree key for a given index.
   * 
   * @param suffix - The index to generate a state tree key for
   * @returns A unique string key for state storage
   */
  private getStateTreeKey(suffix: usize): string {
    return [this.getBaseStatePath(), suffix.toString()].join('_')
  }

  /**
   * Appends an item to the end of the vector.
   * 
   * @param item - The item to append
   */
  push(item: T): void {
    Memory.writeCollection(this.getStateTreeKey(this.length), item)
    this.length += 1
  }

  /**
   * Retrieves the element at the specified index.
   * 
   * @param index - The index of the element to retrieve
   * @returns The element at the index, or null if not found
   */
  get(index: usize): T | null {
    return Memory.readCollection<T>(this.getStateTreeKey(index))
  }

  /**
   * Sets the element at the specified index.
   * 
   * @param index - The index at which to set the element
   * @param item - The item to set at the index
   * @throws {IndexOutOfBoundsError} If the index is out of bounds
   */
  set(index: usize, item: T): void {
    if (index >= this.length) {
      throw new IndexOutOfBoundsError(
        `Index out of bounds: ${index}. Vector length is ${this.length}.`,
      )
    }

    Memory.writeCollection(this.getStateTreeKey(this.length), item)
  }

  /**
   * Removes and returns the last element of the vector.
   * 
   * @returns The last element, or null if the vector is empty
   */
  pop(): T | null {
    if (this.length === 0) {
      return null
    }

    this.length -= 1

    return Memory.deleteCollection<T>(this.getStateTreeKey(this.length))
  }

  /**
   * Executes a callback function for each element in the vector.
   * 
   * @param callbackfn - A function to execute for each element, receiving the value, index, and vector
   * @throws {WeilError} If an element at an index is not set
   */
  forEach(callbackfn: (value: T, index: i32, vec: WeilVec<T>) => void): void {
    for (let i: i32 = 0; i < this.length; i++) {
      const item = this.get(i)
      if (item === null) {
        throw new WeilError(`WeilVec item at index ${i} is not set`)
      } else {
        callbackfn(item, i, this)
      }
    }
  }
}

File: adk/go/weil_go/collections/vec.go
package collections

import (
	"fmt"

	"github.com/weilliptic-public/wadk/adk/go/weil_go/internal/constants"
	"github.com/weilliptic-public/wadk/adk/go/weil_go/runtime"
)

// WeilVec is a vector (dynamic array) collection that can be persisted to the contract's state tree.
// It supports indexed access and dynamic growth.
type WeilVec[T any] struct {
	StateId WeilId `json:"state_id"`
	Len     uint   `json:"len"`
}

// NewWeilVec creates a new empty WeilVec with the given state identifier.
func NewWeilVec[T any](stateId WeilId) *WeilVec[T] {
	return &WeilVec[T]{
		StateId: stateId,
		Len:     0,
	}
}

// BaseStatePath returns the base path for this vector in the state tree.
func (v *WeilVec[T]) BaseStatePath() string {
	return v.StateId.String()
}

// StateTreeKey generates a state tree key for the given index suffix.
func (v *WeilVec[T]) StateTreeKey(suffix uint) string {
	return fmt.Sprintf("%s_%d", v.BaseStatePath(), suffix)
}

// Push appends an item to the end of the vector.
// The vector's length is automatically incremented.
func (v *WeilVec[T]) Push(item *T) {
	index := v.Len
	runtime.WriteCollectionEntry([]byte(v.StateTreeKey(index)), item)
	v.Len += 1
}

// Get retrieves the item at the given index.
// Returns an error if the index is out of bounds.
func (v *WeilVec[T]) Get(index uint) (*T, error) {
	if index >= v.Len {
		return nil, fmt.Errorf("index out of bounds: length of vector is `%d`, index received `%d`", v.Len, index)
	}

	item, err := runtime.ReadCollection[T]([]byte(v.StateTreeKey(index)))

	if err != nil {
		panic(constants.UNREACHABLE)
	}

	return item, nil
}

// Set updates the item at the given index with the provided value.
// Returns an error if the index is out of bounds.
func (v *WeilVec[T]) Set(index uint, item *T) error {
	if index >= v.Len {
		return fmt.Errorf("index out of bounds: length of vector is `%d`, index received `%d`", v.Len, index)
	}

	runtime.WriteCollectionEntry([]byte(v.StateTreeKey(index)), item)

	return nil
}
